osmoutils/binary_search.go

```diff
@@ -93,3 +133,41 @@ func BinarySearch(f func(input sdk.Int) (sdk.Int, error),
 	}
 	return curEstimate, nil
 }
+
+// Binary search BigDec inputs between [lowerbound, upperbound] to a monotonic increasing function f
+// We stop once f(found_input) meets the ErrTolerance constraints.
+// If we perform more than maxIterations (or equivalently lowerbound = upperbound), we return an error.
+func BinarySearchBigDec(f func(input osmomath.BigDec) (osmomath.BigDec, error),
+	lowerbound osmomath.BigDec,
+	upperbound osmomath.BigDec,
+	targetOutput osmomath.BigDec,
+	errTolerance ErrTolerance,
+	maxIterations int,
+) (osmomath.BigDec, error) {
+	// Setup base case of loop
+	curEstimate := lowerbound.Add(upperbound).Quo(osmomath.NewBigDec(2))
+	curOutput, err := f(curEstimate)
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+	curIteration := 0
+	for ; curIteration < maxIterations; curIteration += 1 {
+		compRes := errTolerance.CompareBigDec(curOutput, targetOutput)
+		if compRes > 0 {
+			upperbound = curEstimate
+		} else if compRes < 0 {
+			lowerbound = curEstimate
+		} else {
+			break
```

Comment:
Consider simply returning here

Reply:
This is a good point – after a bit of thought this does seem to be the simplest approach. Updated!